src/CreeDictionary/API/search/eip.py

```diff
@@ -0,0 +1,238 @@
+                words.append(r)
+
+        orig_tags_starting_with_plus: list[str] = []
+        tags_ending_with_plus: list[str] = []
+        for t in self.new_tags:
+            if t.startswith("+"):
+                orig_tags_starting_with_plus.append(t)
+            else:
+                tags_ending_with_plus.append(t)
+
+        results = []
+        for word in words:
+            # This is sometimes mutated
+            tags_starting_with_plus = orig_tags_starting_with_plus[:]
+
+            noun_tags = ""
+            if "N" in word.wordform.inflectional_category:
+                noun_tags = get_noun_tags(word.wordform.inflectional_category)
+                if "+N" in tags_starting_with_plus:
+                    tags_starting_with_plus.remove("+N")
+                if "+Der/Dim" in tags_starting_with_plus:
+                    # noun tags need to be repeated in this case
+                    index = tags_starting_with_plus.index("+Der/Dim")
+                    tags_starting_with_plus.insert(index + 1, noun_tags)
+
+            text = (
+                "".join(tags_ending_with_plus)
+                + word.lemma_wordform.text
+                + noun_tags
+                + "".join(tags_starting_with_plus)
+            )
+
+            generated_wordforms = strict_generator().lookup(text)
+            for w in generated_wordforms:
+                results.append(
+                    _EipResult(original_result=word, inflected_text=w, analysis=text)
+                )
+        return results
+
+
+# e.g., " swim +V+AI+Prt+3Pl"
+PHRASE_ANALYSIS_OUTPUT_RE = re.compile(
+    r"""
+                        \s*             # leading blank space(s) from flag diacritics
+                        (?P<query>.*)
+                        \s
+                        (?P<tags>\+[^\ ]+)
+                    """,
+    re.VERBOSE,
+)
+
+
+class PhraseAnalyzedQuery:
+    """A structured object holding pieces of, and info about, a phrase query.
+
+    >>> PhraseAnalyzedQuery("they swam").filtered_query
+    'swim'
+    >>> PhraseAnalyzedQuery("they swam").has_tags
+    True
+    >>> PhraseAnalyzedQuery("they swam").tags
+    ['+V', '+AI', '+Prt', '+3Pl']
+    >>> PhraseAnalyzedQuery("excellent").has_tags
+    False
+    """
+
+    def __init__(self, query: str, add_verbose_message=None):
+        self.query = query
+        self.has_tags = False
+        self.filtered_query = None
+        self.tags = None
+        phrase_analyses: list[str] = [
+            r.decode("UTF-8") for r in eng_phrase_to_crk_features_fst()[query]
+        ]
+
+        if add_verbose_message:
+            add_verbose_message({"phrase_analyses": phrase_analyses})
+
+        if len(phrase_analyses) != 1:
+            return
+
+        phrase_analysis = phrase_analyses[0]
+        if "+?" in phrase_analysis:
+            return
+
+        match = PHRASE_ANALYSIS_OUTPUT_RE.fullmatch(phrase_analysis)
+        if not match:
+            return
```

Comment:
The new ‘walrus’ operator can make this a little bit nicer:

```suggestion
        if not match := PHRASE_ANALYSIS_OUTPUT_RE.fullmatch(phrase_analysis)
            return
```